flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -2123,6 +2131,92 @@ mlir::Value IntrinsicLibrary::genDprod(mlir::Type resultType,
   return builder.create<mlir::MulFOp>(loc, a, b);
 }
 
+// EOSHIFT
+fir::ExtendedValue
+IntrinsicLibrary::genEoshift(mlir::Type resultType,
+                             llvm::ArrayRef<fir::ExtendedValue> args) {
+  assert(args.size() == 4);
+
+  // Handle required ARRAY argument
+  fir::BoxValue arrayBox = builder.createBox(loc, args[0]);
+  auto array = fir::getBase(arrayBox);
+  auto arrayRank = arrayBox.rank();
+
+  // Create mutable fir.box to be passed to the runtime for the result.
+  auto resultArrayType = builder.getVarLenSeqTy(resultType, arrayRank);
+  auto resultMutableBox =
+      fir::factory::createTempMutableBox(builder, loc, resultArrayType);
+  auto resultIrBox =
+      fir::factory::getMutableIRBox(builder, loc, resultMutableBox);
+
+  // Handle optional BOUNDARY argument
+  mlir::Type arrayEleTy = arrayBox.getEleTy();
+  mlir::Value boundary = nullptr;
+
+  if (isAbsent(args[2])) {
+    if (arrayEleTy.isa<mlir::IntegerType>() ||
+        arrayEleTy.isa<mlir::FloatType>() ||
+        arrayEleTy.isa<fir::LogicalType>()) {
+      // fill in with appropriate number of 0 or .false.
+      auto zero = builder.createIntegerConstant(loc, builder.getIndexType(), 0);
+      auto temp = builder.createTemporary(loc, arrayEleTy);
+      auto cast = builder.createConvert(loc, arrayEleTy, zero);
+      builder.create<fir::StoreOp>(loc, cast, temp);
+      boundary = builder.createBox(loc, temp);
+    } else if (arrayEleTy.isa<fir::CharacterType>()) {
+      // fill in with appropriate number of blanks
+      fir::factory::CharacterExprHelper helper{builder, loc};
+      fir::CharacterType charTy = helper.getCharacterType(arrayEleTy);
+      fir::CharBoxValue temp =
+          helper.createCharacterTemp(arrayEleTy, charTy.getLen());
+      auto zero = builder.createIntegerConstant(loc, builder.getIndexType(), 0);
+      auto len = builder.createIntegerConstant(
+          loc, builder.getCharacterLengthType(), charTy.getLen());
+
+      helper.createPadding(temp, zero, len);
+
+      boundary = builder.createBox(loc, temp);
+    } else
+      fir::emitFatalError(loc, "bad type for EOSHIFT");
+  } else {
+    if (arrayEleTy.isa<mlir::IntegerType>() ||
+        arrayEleTy.isa<mlir::FloatType>() ||
+        arrayEleTy.isa<fir::LogicalType>()) {
+      auto temp = builder.createTemporary(loc, arrayEleTy);
+      auto cast = builder.createConvert(loc, arrayEleTy, fir::getBase(args[2]));
+      builder.create<fir::StoreOp>(loc, cast, temp);
+      boundary = builder.createBox(loc, temp);
+    } else if (arrayEleTy.isa<fir::CharacterType>())
+      boundary = builder.createBox(loc, args[2]);
+    else
+      fir::emitFatalError(loc, "bad type for EOSHIFT");
```

Comment:
I think you can avoid this special handling of having to create a temp in some cases by directly setting `boundary` to `asAddr`, then you should be able to use `builder.createBox(loc, args[2])` in all cases.

This also seems required to cover the cases where boundary may be an array (possible if ARRAY rank is >= 2), in which case the code above will not work.

Reply:
What worked for me was to set the `boundary` argument with the following code:
```
  auto boundary = isAbsent(args, 2)
                      ? builder.create<fir::AbsentOp>(
                            loc, fir::BoxType::get(builder.getNoneType()))
                      : builder.createBox(loc, args[2]);
```

Reply:
Thanks for the asAddr catch!  And yes, I agree this is a much better way to handle things.  I have made the necessary updates.